src/platform/ESP32/NetworkCommissioningWiFiDriver.cpp

```diff
@@ -0,0 +1,327 @@
+
+    err               = ConnectWiFiNetwork(reinterpret_cast<const char *>(mStagingNetwork.ssid), mStagingNetwork.ssidLen,
+                             reinterpret_cast<const char *>(mStagingNetwork.credentials), mStagingNetwork.credentialsLen);
+    mpConnectCallback = callback;
+exit:
+    if (err != CHIP_NO_ERROR)
+    {
+        networkingStatus = Status::kUnknownError;
+    }
+    if (networkingStatus != Status::kSuccess)
+    {
+        ChipLogError(NetworkProvisioning, "Failed to connect to WiFi network:%s", chip::ErrorStr(err));
+        mpConnectCallback = nullptr;
+        callback->OnResult(networkingStatus, CharSpan(), 0);
+    }
+}
+
+CHIP_ERROR ESPWiFiDriver::StartScanWiFiNetworks(ByteSpan ssid)
+{
+    esp_err_t err = ESP_OK;
+    if (ssid.data())
+    {
+        wifi_scan_config_t scan_config = { 0 };
+        memset(WiFiSSIDStr, 0, sizeof(WiFiSSIDStr));
+        memcpy(WiFiSSIDStr, ssid.data(), ssid.size());
+        scan_config.ssid = WiFiSSIDStr;
+        err              = esp_wifi_scan_start(&scan_config, false);
+    }
+    else
+    {
+        err = esp_wifi_scan_start(NULL, false);
+    }
+    if (err != ESP_OK)
+    {
+        return chip::DeviceLayer::Internal::ESP32Utils::MapError(err);
+    }
+    return CHIP_NO_ERROR;
+}
+
+bool GetWiFiScanRespFromAPRecord(wifi_ap_record_t * ap_record, WiFiScanResponse & scanResponse)
+{
+    scanResponse.security = ap_record->authmode;
+    scanResponse.ssidLen  = strnlen((const char *) ap_record->ssid, DeviceLayer::Internal::kMaxWiFiSSIDLength);
+    memcpy((char *) scanResponse.ssid, ap_record->ssid, scanResponse.ssidLen);
+    memcpy(scanResponse.bssid, ap_record->bssid, 6);
+    scanResponse.channel  = ap_record->primary;
+    scanResponse.wiFiBand = DeviceLayer::NetworkCommissioning::WiFiBand::k2g4; // ESP32 platform only support 2.4G Hz WiFi Band
+    scanResponse.rssi     = ap_record->rssi;
+    return true;
+}
+
+void ESPWiFiDriver::OnScanWiFiNetworkDone()
+{
+    uint16_t ap_number;
+    esp_wifi_scan_get_ap_num(&ap_number);
+    if (!ap_number)
+    {
+        ChipLogProgress(DeviceLayer, "No AP found");
+        if (mpScanCallback != nullptr)
+        {
+            mpScanCallback->OnFinished(Status::kSuccess, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    wifi_ap_record_t * ap_list_buffer;
+    ap_list_buffer = (wifi_ap_record_t *) malloc(ap_number * sizeof(wifi_ap_record_t));
+    if (ap_list_buffer == NULL)
+    {
+        ChipLogError(DeviceLayer, "can't malloc memory for ap_list_buffer");
+        if (mpScanCallback)
+        {
+            mpScanCallback->OnFinished(Status::kUnknownError, CharSpan(), nullptr);
+            mpScanCallback = nullptr;
+        }
+        return;
+    }
+    if (esp_wifi_scan_get_ap_records(&ap_number, (wifi_ap_record_t *) ap_list_buffer) == ESP_OK)
+    {
+        DeviceLayer::SystemLayer().ScheduleLambda([ap_number, ap_list_buffer]() {
```

Comment:
ScheduleLambda returns a CHIP_ERROR, so it may fail. That case needs to also be considered when freeing up the buffer.

Reply:
Please handle ScheduleLambda returning failure (log and free the buffer).

Reply:
Done

Reply:
I expect a `err = DeviceLayer::SystemLayer().ScheduleLambda` or some `if (CHIP_NO_ERROR != DeviceLayer::SystemLayer().ScheduleLambda` or similar in here.

This comment was marked resolved twice, yet the "free the buffer if Schedule lambda returns an error" does not seem handled.

This handling of freeing seems very error prone to me. Consider:

```
std::unique_ptr<wifi_ap_record_t[]> buffer_ptr(new wifi_ap_record_t[ap_number]);
if (!buffer_ptr) {
   // ...
}
// ....
    wifi_ap_record_t * ap_list_buffer = buffer_ptr.get();
    if (CHIP_NO_ERROR == DeviceLayer::SystemLayer().ScheduleLambda([ap_number, ap_list_buffer] {
        std::unique_ptr<wifi_ap_record_t[]> auto_free(ap_list_buffer);
        // ...
        )) {
       buffer_ptr.release();  // lambda takes ownership and frees the buffer.
    }
```